ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl.hs

```diff
@@ -249,24 +248,29 @@ putBlockImpl :: forall m blockId. (IOLike m, Ord blockId)
              -> BlockInfo blockId
              -> BS.Builder
              -> m ()
-putBlockImpl env@VolatileDBEnv{ maxBlocksPerFile, tracer }
+putBlockImpl env@VolatileDBEnv{ maxBlocksPerFile, tracer, prefixLen }
              blockInfo@BlockInfo { bbid, bslot, bpreBid }
              builder =
     appendOpenState env $ \hasFS -> do
       OpenState { currentRevMap, currentWriteHandle } <- get
       if Map.member bbid currentRevMap then
         lift $ lift $ traceWith tracer $ BlockAlreadyHere bbid
       else do
-        bytesWritten <- lift $ lift $ hPut hasFS currentWriteHandle builder
+        -- The builder will likely consist of a single chunk as the annotation
+        -- of the block is used here, which is strict. Related: #1215.
+        let bytes = BS.toLazyByteString builder
+        bytesWritten <- lift $ lift $ hPutAll hasFS currentWriteHandle bytes
         fileIsFull <- state $ updateStateAfterWrite bytesWritten
+                                                    (takePrefix prefixLen bytes)
```

Comment:
I wonder if we should do this the other way around. Suppose the `Builder` does _not_ correspond to something already in memory. The way you do it now first writes it to disk, forcing the entire thing into memory (because you're holding on to the start), and _then_ gets the first few bytes. if you did it the other way around, you'd only force the first chunk in memory, and GC can start to collect stuff when the file is being written. (Would need to be sure of course that we also force the short bytestring.)